### PR TITLE
Mldb 1255 Fix calling dataset.rowName in N-way joins 

### PR DESCRIPTION
**Pull request by steadtler**
And also replace empty rowname entries with "null" in outer joins.

I wasnt sure is getOriginalRowName should be part of the dataset interface or the MatrixView interface - I picked the dataset interface because I thought MatrixView and ColumnIndex were mainly meant as pure interfaces, while getOriginalRowName should only be overriden by the JoinDataset. But Im open to suggestions.

I upgraded the PR to include the fixes to MLDB-1305 since no one had reviewed it yet.

### CODE REVIEW

File: testing/MLDB-1305_rowNames_join.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+
+# This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.
```
Comment (guyd) [inline testing/MLDB-1305_rowNames_join.py]: 2015 -> 2016

File: testing/MLDB-1305_rowNames_join.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+
+res = mldb.perform("GET", "/v1/query", [["q", """
+    SELECT dataset1.rowName(), dataset2.rowName()
+    FROM dataset1
+    JOIN dataset2 ON dataset2.rowName() = dataset1.rowName()
+    """]])
+
+mldb.log(res)
+assert res["statusCode"] != 404
```
Comment (guyd) [inline testing/MLDB-1305_rowNames_join.py]: This does not test much.  Can we check for the body as well?  And why 404?

File: types/string.cc [hunk 3]
```diff
@@ -518,6 +518,32 @@ removePrefix(const Utf8String & prefix)
     return false;
 }
 
+bool
+Utf8String::
+removeSuffix(const Utf8String & suffix)
+{
+   if (suffix.empty())
```
Comment (guyd) [inline types/string.cc]: indentation problem

File: testing/MLDB-1305_rowNames_join.py (hunk 2 above)
Comment (steadtler) [inline testing/MLDB-1305_rowNames_join.py]: Francois wrote the test. I could check some of the output...

**Comment by guyd** [comment]
good test coverage +1

**Pull request merged**